work_queue/src/work_queue.c

```diff
@@ -827,6 +850,72 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
+static void remove_factory(struct work_queue_factory_info *f)
+{
+	free(f->name);
+	free(f);
+}
+
+static void update_factory(struct work_queue *q, struct jx *j)
+{
+	const char *name = jx_lookup_string(j, "factory_name");
+	struct work_queue_factory_info *f;
+
+	if ( !(f = hash_table_lookup(q->factory_table, name)) ) {
+		// Create a new factory
+		debug(D_WQ, "new factory %s detected", name);
+		f = malloc(sizeof(*f));
+		if (!f) {
+			debug(D_NOTICE, "Cannot allocate memory for factory %s.", name);
+			return;
+		}
+		memset(f, 0, sizeof(*f));
+	}
+
+	f->name        = xxstrdup(name);
+	f->num_workers = 0;
+	hash_table_insert(q->worker_table, f->name, f);
+
+	return;
+}
+
+void update_read_catalog_factory(struct work_queue *q, time_t stoptime) {
+	struct catalog_query *cq;
+	struct jx *jexpr = NULL;
+	struct jx *j;
+
+	// Iterate through factory_table to create a query filter.
+	int first_name = 1;
+	char filter[BUFSIZ];
+	buffer_init((buffer_t *) filter);
+	char **factory_name = NULL;
+	struct work_queue_factory_info **f = NULL;
+	sprintf(filter, "type == \"wq_factory\" && (");
```

Comment:
Don't use sprintf. Instead see `dttools/src/buffer.h` on how to append to a buffer.

Reply:
Here I used `NULL` as the `size` in `buffer_tolstring`.